### ISSUE TITLE
Issues with flyspell mode in code blocks

### ISSUE THREAD

**Issue by phst** [jrblevin/markdown-mode#311]
Code in code blocks is incorrectly treated as spelling error by Flyspell.

## Expected Behavior

Code in code blocks is never treated as spelling error.

## Actual Behavior

Code in code blocks is occasionally treated as spelling error (see reproduction steps below).

## Steps to Reproduce

1. Start Emacs and load markdown-mode.
1. Visit a nonexisting Markdown file.
1. Enable flyspell-mode (<kbd>M-x flyspell-mode</kbd>)
1. Insert a code block with nonexisting words, such as
   ~~~markdown
   ```aaa
   bbb
   ```
   ~~~

"aaa" and "bbb" will now get the red Flyspell underline because they are spelling errors, but Flyspell should ignore code. Interestingly, `markdown-flyspell-check-word-p` returns nil inside the code block, i.e. it's detected correctly. Saving and then reverting the buffer using <kbd>M-x revert-buffer</kbd> doesn't help. Only turning Flyspell mode off and on again (by hitting <kbd>M-x flyspell-mode</kbd> twice) works.

## Software Versions

- Markdown Mode: From Git on 2018-01-24, version 2.4-dev
- Emacs: GNU Emacs 27.0.50 (build 5, x86_64-pc-linux-gnu, GTK+ Version 3.22.24) of 2018-01-24
- OS: Debian testing

**Comment by jrblevin** [comment]
Thanks for the report, however, I can't reproduce this. I tried first with my usual local configuration, which enables flyspell during my markdown-mode hook. Then I tried starting from `emacs -Q -l markdown-mode.el`. Both times, the code block was ignored.  One difference is that I'm using Emacs 26.0.91, but that seems unlikely to make a difference.

**Comment by saf-dmitry** [comment]
I would assume (maybe I'm wrong), that when you type the code block symbol-by-symbol (instead copying and pasting is as a whole), then until the closing ticks have matched the opening ticks, it is _still not recognized as a code block_ by Font Lock, and the corresponding text properties (e.g., `markdown-pre`) are not set. Flyspell puts its overlays over the "misspelled" words and those overlays will be removed by Flyspell when you toggle it off.

My suggestion is to define a function like
```elisp
(defun markdown-remove-flyspell-overlays-in (begin end)
  "Remove Flyspell overlays in region between BEGIN and END."
  (and (bound-and-true-p flyspell-mode)
       (fboundp 'flyspell-delete-region-overlays)
       (flyspell-delete-region-overlays begin end)))
```
This function should be plugged into the Font Lock machinery and called after a code block (or other constructs, which should to be spell-checked) is fontified with the boundaries of this block.

**Comment by jrblevin** [comment]
I yanked the code block in, so it was recognized immediately, so this is probably why I couldn’t reproduce. I’ll double check. Thanks Dmitry for the function.

**Comment by saf-dmitry** [comment]
I use a similar function in the [TaskPaper mode](https://github.com/saf-dmitry/taskpaper-mode)  (sorry for the shameless plug) to remove Flyspell overlays from the hashtags after their fontification. It does the job without any fuss.

**Comment by syohex** [comment]
I have merged #501. Please check latest version.
